chainstate/src/detail/chainstateref.rs

```diff
@@ -422,46 +427,393 @@ impl<'a, S: BlockchainStorageRead, O: OrphanBlocks> ChainstateRef<'a, S, O> {
+                    ));
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn get_total_coins_in_outputs(
+        &self,
+        tx: &Transaction,
+        block_id: Id<Block>,
+    ) -> Result<Amount, CheckBlockTransactionsError> {
+        let res: Amount = tx
+            .outputs()
+            .iter()
+            .filter_map(|output| match output.value() {
+                OutputValue::Coin(coin) => Some(coin),
+                OutputValue::Asset(_) => None,
+            })
+            .try_fold(Amount::from_atoms(0), |acc, coin| acc + *coin)
+            .ok_or_else(|| {
+                CheckBlockTransactionsError::CoinOrAssetOverflow(tx.get_id(), block_id)
+            })?;
+        Ok(res)
+    }
+
+    fn get_total_coins_in_inputs(
+        &self,
+        tx: &Transaction,
+        block_id: Id<Block>,
+    ) -> Result<Amount, CheckBlockTransactionsError> {
+        tx.inputs()
+            .iter()
+            .filter_map(|input| {
+                let prev_tx = self.fetch(input.outpoint()).ok()?;
+                let output = prev_tx.outputs().get(input.outpoint().output_index() as usize)?;
+                Some(output.value().clone())
+            })
+            .filter_map(|output_value| match output_value {
+                OutputValue::Coin(coin) => Some(coin),
+                OutputValue::Asset(_) => None,
+            })
+            .try_fold(Amount::from_atoms(0), |acc, coin| {
+                (acc + coin).ok_or_else(|| {
+                    CheckBlockTransactionsError::CoinOrAssetOverflow(tx.get_id(), block_id.clone())
+                })
+            })
+    }
+
+    fn is_issuance_fee_enough(coins_in_inputs: Amount, coins_in_outputs: Amount) -> bool {
+        (coins_in_inputs - coins_in_outputs).unwrap_or(Amount::from_atoms(0))
+            > TOKEN_MIN_ISSUANCE_FEE
+    }
+
+    fn check_burn_data(
+        &self,
+        tx: &Transaction,
+        block_id: &Id<Block>,
+        burn_token_id: &TokenId,
+        amount_to_burn: &Amount,
+    ) -> Result<(), CheckBlockTransactionsError> {
+        // Collect token inputs
+        let total_value_tokens = self.get_total_value_tokens(tx, block_id)?;
+
+        // Is token exist in inputs?
+        let (_, origin_amount) = total_value_tokens
+            .iter()
+            .find(|&(origin_token_id, _)| origin_token_id == burn_token_id)
+            .ok_or_else(|| {
+                CheckBlockTransactionsError::NoTokenInInputs(tx.get_id(), block_id.clone())
+            })?;
+
+        // Check amount
+        if origin_amount < amount_to_burn {
+            return Err(CheckBlockTransactionsError::InsuffienceTokenValueInInputs(
+                tx.get_id(),
+                block_id.clone(),
+            ));
+        }
+        if amount_to_burn == &Amount::from_atoms(0) {
+            return Err(CheckBlockTransactionsError::BurnZeroTokens(
+                tx.get_id(),
+                block_id.clone(),
+            ));
+        }
```

Comment:
`ensure!()`?

Reply:
Thanks!

Reply:
Unfortunately, on some places like `check_issue_data` it's inconvenient, because cause errors to borrow-checker. I think, there is better to use `if` and `return` rather a bunch of `clone` on `tx_id` and `block_id`

Reply:
In other places, I started to use `ensure!`. If you are happy with that, I resolve conversation